In [41]:
import squidpy as sq
import seaborn as sns
import shout
import pandas as pd
from IPython.display import set_matplotlib_formats
import random
import scanpy as sc

In [2]:
mibitof = sq.datasets.mibitof()

In [3]:
mibitof

AnnData object with n_obs × n_vars = 3309 × 36
    obs: 'row_num', 'point', 'cell_id', 'X1', 'center_rowcoord', 'center_colcoord', 'cell_size', 'category', 'donor', 'Cluster', 'batch', 'library_id'
    var: 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2'
    uns: 'Cluster_colors', 'batch_colors', 'neighbors', 'spatial', 'umap'
    obsm: 'X_scanorama', 'X_umap', 'spatial'
    obsp: 'connectivities', 'distances'

In [30]:
mibitof_df=mibitof.to_df()

In [156]:
indices=mibitof.to_df().index
indices=indices.to_list()

In [34]:
len(indices)

3309

In [37]:
mibitof_df=mibitof_df.reset_index(drop=True)

In [38]:
mibitof_df

ASCT2     ATP5A     CD11c      CD14       CD3      CD31      CD36   
0    -0.048445  0.137012 -0.086196 -0.080735 -0.113008 -0.078852 -0.046787  \
1    -0.103411  0.106259 -0.077513 -0.067131 -0.108321 -0.072423 -0.047707   
2    -0.058778 -0.183549 -0.057169 -0.252341 -0.074428 -0.043755 -0.053108   
3    -0.063266 -0.217120 -0.010480 -0.018193 -0.012152 -0.066393 -0.016208   
4    -0.195802  0.225289 -0.038972 -0.017998 -0.078372 -0.053338 -0.055921   
...        ...       ...       ...       ...       ...       ...       ...   
3304  0.021182  0.045632 -0.122180 -0.159719 -0.107684  0.165514 -0.070085   
3305 -0.132899 -0.285047 -0.028203  0.133393 -0.082585 -0.069386 -0.066176   
3306 -0.145032 -0.038225  0.079962  0.143460 -0.112061 -0.083390 -0.069760   
3307 -0.110586 -0.288419 -0.096880  0.083653 -0.094524 -0.061432 -0.060936   
3308 -0.094266 -0.198548 -0.063249 -0.160422 -0.056381 -0.039209 -0.048076   

          CD39       CD4      CD45  ...     NRF2p  NaKATPase       PD1   
0    -0.141266 -0.085507 -0.248544  ... -0.054893   0.065361 -0.083144  \
1    -0.083563 -0.113500 -0.208693  ...  0.009360   0.025300 -0.081925   
2    -0.119548 -0.157841 -0.049226  ... -0.135845  -0.155732 -0.066608   
3    -0.110662 -0.012966  0.106709  ... -0.156271  -0.121936 -0.065462   
4    -0.064804 -0.047809  0.046930  ...  0.029879  -0.219233 -0.077997   
...        ...       ...       ...  ...       ...        ...       ...   
3304 -0.132370 -0.135155 -0.189867  ...  0.067894  -0.108838 -0.106829   
3305 -0.132192 -0.039072  0.073765  ... -0.234625  -0.225610 -0.094546   
3306  0.272488  0.175698  0.181223  ...  0.022217  -0.091115 -0.102767   
3307 -0.264610 -0.170266  0.018944  ... -0.049546  -0.020856 -0.089480   
3308 -0.226687 -0.196662 -0.064409  ... -0.151842  -0.205693 -0.057455   

          PKM2       S6p      SDHA       SMA     VDAC1      XBP1  vimentin  
0    -0.184737 -0.038722  0.093941 -0.322605  0.080074 -0.000947 -0.321359  
1    -0.240841 -0.033127  0.071607 -0.281874  0.034381 -0.027246 -0.269943  
2    -0.133573 -0.206725 -0.176896  0.091195 -0.258185 -0.045451 -0.104695  
3     0.053975 -0.090116 -0.172565  0.086739 -0.233419 -0.040213  0.028672  
4    -0.154873 -0.078090  0.092791  0.106398 -0.155387 -0.040580  0.011732  
...        ...       ...       ...       ...       ...       ...       ...  
3304 -0.059154  0.036202 -0.091055  0.239040  0.084519  0.175693  0.180371  
3305  0.039629 -0.146379 -0.172444  0.174762 -0.155889  0.153453  0.012337  
3306  0.153053  0.181595 -0.008561 -0.027603 -0.293268 -0.255045  0.253582  
3307 -0.041875 -0.087478 -0.175069  0.079871 -0.381522 -0.116259  0.080427  
3308 -0.012602 -0.081220 -0.205746  0.164505 -0.353435 -0.025183 -0.055012  

[3309 rows x 36 columns]

AnnData object with n_obs × n_vars = 3309 × 36
    obs: 'row_num', 'point', 'cell_id', 'X1', 'center_rowcoord', 'center_colcoord', 'cell_size', 'category', 'donor', 'Cluster', 'batch', 'library_id'
    var: 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2'
    uns: 'Cluster_colors', 'batch_colors', 'neighbors', 'spatial', 'umap'
    obsm: 'X_scanorama', 'X_umap', 'spatial'
    obsp: 'connectivities', 'distances'

In [184]:
# --- Imports: ---
import time
# ---- Code: ----
# shout_scores={}
# n=[5, 10, 50, 100, 200, 500, 1000]
n=[50, 100, 150]
cluster_key='Cluster'
indices_range=range(len(indices))

LOCAL_ENTROPY_2={}
LOCAL_HOMOPHILY_2={}
EGOPHILY_2={}


for i in n:
    LOCAL_ENTROPY_2[i]=[]
    LOCAL_HOMOPHILY_2[i]=[]
    EGOPHILY_2[i]=[]
    # ---
    local_entropy_2={}
    local_homophily_2={}
    egophily_2={}
    # ---
    for j in range(100):
        # print('Iteration '+str(_count_+1)+', i= '+str(i))
        # ---
        rs=random.sample(indices_range, i)
        # print(mibitof_df.loc[rs])
        mibitof_rand=sc.AnnData(mibitof_df.iloc[rs])
        # ---
        _mibitof_obs_df_=pd.DataFrame(mibitof.obs)
        for j in mibitof.obs.columns:
            mibitof_obs_df=_mibitof_obs_df_[j].iloc[rs]
            mibitof_rand.obs[j]=np.array(mibitof_obs_df)
        # ---
        mibitof_rand.var=mibitof.var
        # ---
        mibitof_rand.uns=mibitof.uns
        # ---
        _mibitof_obsm_df_=pd.DataFrame(mibitof.obsm)
        for j in mibitof.obsm:
            mibitof_obsm_df=pd.DataFrame(mibitof.obsm[j]).iloc[rs]
            mibitof_rand.obsm[j]=np.array(mibitof_obsm_df)
        # ---
        # _mibitof_obsp_df_=pd.DataFrame(mibitof.obsp)
        # for j in mibitof.obsp:
            # mibitof_obsp_df=pd.DataFrame(mibitof.obsp[j]).iloc[rs]
            # mibitof_rand.obsp[j]=np.array(mibitof_obsp_df)
        # ---
        scores_mibitof = shout.all_scores(mibitof_rand, cluster_key, radius=2)
        # end = time.time()
        # ============================
        
        scores=[k for k in pd.DataFrame(mibitof_rand.obs['local_entropy_2']).to_dict()['local_entropy_2'].values()]
        celltypes=[k for k in pd.DataFrame(mibitof_rand.obs[cluster_key]).to_dict()[cluster_key].values()]
        count=-1
        for k in celltypes:
            count+=1
            try:
                local_entropy_2[k].append(scores[count])
            except:
                local_entropy_2[k]=[]
                local_entropy_2[k].append(scores[count])
        LOCAL_ENTROPY_2[i].append(local_entropy_2)
        # ============================
        
        scores=[k for k in pd.DataFrame(mibitof_rand.obs['local_homophily_2']).to_dict()['local_homophily_2'].values()]
        celltypes=[k for k in pd.DataFrame(mibitof_rand.obs[cluster_key]).to_dict()[cluster_key].values()]
        count=-1
        for k in celltypes:
            count+=1
            try:
                local_homophily_2[k].append(scores[count])
            except:
                local_homophily_2[k]=[]
                local_homophily_2[k].append(scores[count])
        LOCAL_HOMOPHILY_2[i].append(local_homophily_2)
        # ============================
        
        scores=[k for k in pd.DataFrame(mibitof_rand.obs['egophily_2']).to_dict()['egophily_2'].values()]
        celltypes=[k for k in pd.DataFrame(mibitof_rand.obs[cluster_key]).to_dict()[cluster_key].values()]
        count=-1
        for k in celltypes:
            count+=1
            try:
                egophily_2[k].append(scores[count])
            except:
                egophily_2[k]=[]
                egophily_2[k].append(scores[count])
        EGOPHILY_2[i].append(egophily_2)
        # ============================

/Users/surya/anaconda3/envs/spatial_proteomics_cloned_xai/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.
  warnings.warn("Transforming to str index.", ImplicitModificationWarning)
/Users/surya/anaconda3/envs/spatial_proteomics_cloned_xai/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.
  warnings.warn("Transforming to str index.", ImplicitModificationWarning)
/Users/surya/anaconda3/envs/spatial_proteomics_cloned_xai/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.
  warnings.warn("Transforming to str index.", ImplicitModificationWarning)
/Users/surya/anaconda3/envs/spatial_proteomics_cloned_xai/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.
  warnings.warn("Transforming to str index.", ImplicitModificationWarning)
/Users/s

In [183]:
LOCAL_ENTROPY_2

{50: [{'Tcell_CD4': [0.9384007999934956,
    0.8964154761697195,
    0.8278195311147832,
    0.8410771429325403,
    0.8590003631325264,
    0.886709917544419,
    0.8647059638770619,
    0.8424308615562998,
    0.8414658234492332,
    0.6798615607440995,
    0.8089289233119883,
    0.8378766015586174,
    0.8901366706634252,
    0.8319238647621109,
    0.7903690218502718,
    0.8176226787652126,
    0.7678541267769687,
    0.8038045864457504,
    0.8799477115559928,
    0.8790219611786495,
    0.7664379815146259,
    0.8637438807377282,
    0.8042151889634931,
    0.8759661622068269,
    0.8440582219360193,
    0.8566046578461299,
    0.8847517710716762,
    0.7842485936239566,
    0.8134079763139508,
    0.8172527613754791,
    0.8045998548685601,
    0.873784204537674,
    0.8302469201035917,
    0.8622959469986046,
    0.8647059638770619,
    0.8481771013829448,
    0.8569835314848895],
   'Epithelial': [0.7690628587208895,
    0.9090723338208275,
    0.8054999283272369,
    0.7322

In [128]:
shout.all_scores(mibitof_rand, cluster_key='Cluster', radius=2)

In [129]:
mibitof_rand

AnnData object with n_obs × n_vars = 150 × 36
    obs: 'row_num', 'point', 'cell_id', 'X1', 'center_rowcoord', 'center_colcoord', 'cell_size', 'category', 'donor', 'Cluster', 'batch', 'library_id', 'local_entropy_2', 'local_homophily_2', 'egophily_2'
    var: 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2'
    uns: 'Cluster_colors', 'batch_colors', 'neighbors', 'spatial', 'umap', 'global_entropy', 'global_homophily'
    obsm: 'X_scanorama', 'X_umap', 'spatial'

In [173]:
local_entropy_2

{'Fibroblast': [0.8311803634512183,
  0.8278195311147832,
  0.8701924144438808,
  0.9202222757438662,
  0.8326925155412138,
  0.7945310730498214,
  0.7840723338208275,
  0.8622731037219674,
  0.8045304344395722,
  0.9465918465485027,
  0.6991772419328298,
  0.8075265872808305,
  0.8018796874098556,
  0.9240618201705576,
  0.8345672320191161,
  0.8348978069430468,
  0.8249639059499646,
  0.6583125004006424,
  0.8134079763139508],
 'Epithelial': [0.7994259655786021,
  0.8654479953199968,
  0.49888025406228603,
  0.7427237314469578,
  0.7189265498156798,
  0.7625945733136903,
  0.7103924194125386,
  0.8593958348795886,
  0.7274095729712671,
  0.8170466374027375,
  0.7514825263240316,
  0.7463659726629288,
  0.708871490603394,
  0.8075265872808307,
  0.7256360563895261,
  0.671044155056913,
  0.7135354866336909,
  0.7711532500572749,
  0.6430309884992668,
  0.8416483250282373,
  0.7809125495378889,
  0.8052204900639212,
  0.7929973289666877,
  0.681370139249467,
  0.7078262241899623,
  0.7

In [131]:
a=shout.all_scores(mibitof_rand, cluster_key='Cluster', radius=2)

In [154]:
np.array(pd.DataFrame(mibitof_rand.obs['local_entropy_2']).to_dict()['local_entropy_2'].values())

array(dict_values([0.8311803634512183, 0.7994259655786021, 0.8654479953199968, 0.6879550406502847, 0.49888025406228603, 0.6226551194132369, 0.8278195311147832, 0.8695649898542807, 0.7427237314469578, 0.7189265498156798, 0.7718892921256691, 0.8701924144438808, 0.7842485936239566, 0.8347171654595357, 0.8937587898675271, 0.7618668463207694, 0.8700211802923575, 0.7625945733136903, 0.8569835314848895, 0.9202222757438662, 0.860337800525862, 0.8216898206052793, 0.8129146401339158, 0.7103924194125386, 0.8805347226228645, 0.8331814352027106, 0.8593958348795886, 0.8326925155412138, 0.7522168043752706, 0.7274095729712671, 0.7987468750600963, 0.8170466374027375, 0.7945310730498214, 0.8219857736497249, 0.8022221055802298, 0.7514825263240316, 0.6070106527112737, 0.7025855388402041, 0.8622959469986046, 0.86292639519725, 0.8024766742935392, 0.850944053893068, 0.8176065103715944, 0.800352774164792, 0.7013033034274548, 0.8832648206811903, 0.7463659726629288, 0.8995534868003716, 0.708871490603394, 0.8348

In [133]:
type(a)

NoneType

In [149]:
pd.DataFrame(mibitof_rand.obs[cluster_key]).to_dict()[cluster_key].values()

dict_values(['Fibroblast', 'Epithelial', 'Epithelial', 'Tcell_CD4', 'Epithelial', 'Myeloid_CD68', 'Fibroblast', 'Tcell_CD4', 'Epithelial', 'Epithelial', 'Tcell_CD4', 'Fibroblast', 'Imm_other', 'Tcell_CD4', 'Tcell_CD4', 'Tcell_CD8', 'Myeloid_CD11c', 'Epithelial', 'Tcell_CD4', 'Fibroblast', 'Myeloid_CD68', 'Tcell_CD4', 'Tcell_CD8', 'Epithelial', 'Imm_other', 'Tcell_CD8', 'Epithelial', 'Fibroblast', 'Imm_other', 'Epithelial', 'Imm_other', 'Epithelial', 'Fibroblast', 'Tcell_CD4', 'Myeloid_CD68', 'Epithelial', 'Imm_other', 'Tcell_CD4', 'Myeloid_CD11c', 'Myeloid_CD11c', 'Myeloid_CD11c', 'Tcell_CD4', 'Imm_other', 'Tcell_CD4', 'Imm_other', 'Tcell_CD4', 'Epithelial', 'Imm_other', 'Epithelial', 'Myeloid_CD68', 'Tcell_CD4', 'Tcell_CD8', 'Myeloid_CD68', 'Epithelial', 'Myeloid_CD68', 'Epithelial', 'Endothelial', 'Tcell_CD4', 'Imm_other', 'Fibroblast', 'Myeloid_CD68', 'Tcell_CD4', 'Fibroblast', 'Epithelial', 'Myeloid_CD68', 'Tcell_CD8', 'Fibroblast', 'Epithelial', 'Myeloid_CD68', 'Tcell_CD8', 'Imm_o

In [135]:
type(mibitof_rand.obs['Cluster'])

pandas.core.series.Series